Enable checking FIPS compliance on container images with imageextractor and fipschecker

everjing opened the pull request:
Add image reference extractor to extract images reference from component config files and add a fips checker to check their fips compliance.

Commit:
add imageextractor package for image reference extraction from AST

- Add ExtractImages() function to traverse AST and extract container image references
- Support complex nested structures (arrays, objects, Kubernetes YAML)
- Extract images with precise AST path mapping for result injection
- Include comprehensive test suite with AST path verification

Commit:
add fipschecker package for FIPS compliance validation

- Add FIPSClient for checking FIPS compliance of container images extracted from each config file
- Support real Docker image pulling and Go binary analysis
- Include host FIPS environment validation with OpenSSL checking
- Add file-level compliance aggregation (file compliant only if all images compliant)
- Comprehensive integration tests with real Microsoft container images
- Update go.mod dependencies and rego query integration

stanleyt10 commented on sg/internal/engine/rego_query.go:
```diff
@@ -35,6 +37,56 @@ func loadSource(source source.Source, shouldParseArmTemplateDefaults bool) ([]lo
 			armtemplateparser.ParseArmTemplateDefaults(t)
 		}
 
+		// Extract images from the AST
```
Can you pull this logic into a separate file, then call oneliner like extractor.ExtractImages(t) similar to ParseArmTemplateDefaults. 

So we can seperate logic

stanleyt10 commented on sg/internal/engine/rego_query.go:
```diff
@@ -20,7 +22,7 @@ type loadedConfiguration struct {
 }
 
-func loadSource(source source.Source, shouldParseArmTemplateDefaults bool) ([]loadedConfiguration, error) {
+func loadSource(source source.Source, shouldParseArmTemplateDefaults bool, enableFIPSCheck bool, fipsScriptPath string) ([]loadedConfiguration, error) {
```
Can you pull all options into a ...opts and do an apply for each one. In the future if we add more flags this will quickly get overloaded.

stanleyt10 replied:
Also can you add all fips logic behind a feature flag. We don't want to change the default behavior of sg for everyone

stanleyt10 commented on sg/internal/engine/rego_query.go:
```diff
@@ -35,6 +37,56 @@ func loadSource(source source.Source, shouldParseArmTemplateDefaults bool) ([]lo
+
+			// Run FIPS check if enabled
+			if enableFIPSCheck {
+				fipsClient := fipschecker.NewFIPSClient(fipsScriptPath)
```
Can you create client at top level, or in another file ?

stanleyt10 commented on sg/internal/engine/rego_query.go:
```diff
@@ -35,6 +37,56 @@ func loadSource(source source.Source, shouldParseArmTemplateDefaults bool) ([]lo
+							if result.IsCompliant {
+								status = "✅ COMPLIANT"
+							}
+							fmt.Printf("  FIPS: %s -> %s\n", result.Image, status)
```
Adding prints here will mess with our sg-project output parsing later. Can you either hide these behind a debug flag if you need them or preferably add code to sg-project to parse the json and print output there

stanleyt10 commented on sg/internal/imageextractor/extractor.go:
```diff
@@ -0,0 +1,97 @@
+
+				// If this is an "image" field and value is a string, record it
+				// Skip FIPS compliance injected fields to avoid duplicate extraction
+				if key == "image" && !containsFIPSCompliancePath(pathStack) {
```
Can you review this de-duping logic. Its checking if fips_compliance is in the pathPart but never adds is to parsed paths. Can just using set to keep track of parsed images & de-dupe

stanleyt10 commented on sg/internal/fipschecker/client.go:
```diff
@@ -0,0 +1,329 @@
+
+// AggregateResultsByFile takes image-level FIPS results and creates file-level results
+// A file is compliant only if ALL images in the file are compliant
+func (c *FIPSClient) AggregateResultsByFile(fipsResults map[string]FIPSResult) map[string]FileFIPSResult {
```
Is this used anywhere ?

stanleyt10 commented on sg/internal/fipschecker/client.go:
```diff
@@ -0,0 +1,329 @@
+}
+
+// CheckMultipleImages runs FIPS check on multiple images
+func (c *FIPSClient) CheckMultipleImages(images []imageextractor.ImageReference) (map[string]FIPSResult, error) {
```
Can we check images in parallel otherwise will be very slow to pull every single image. 
Also can speed up with caching results of shared images.

stanleyt10 commented on sg/internal/fipschecker/client.go:
```diff
@@ -0,0 +1,329 @@
+}
+
+// checkWithGoAPI uses the fips-check Go API to scan container image
+func (c *FIPSClient) checkWithGoAPI(imageRef imageextractor.ImageReference) (FIPSResult, error) {
```
can we use https://github.com/[bahe-msft/fips-check](https://github.com/bahe-msft/fips-check/tree/main)/tree/main

Commit:
refactor: replace Go SDK with fips-check shell script integration

Major changes:
- Replace complex Go SDK approach with simple shell script execution
- Remove obsolete packages: binarychecker, opensslsetup, cmd/fips-checker
- Simplify FIPSClient to execute ./build-and-check.sh directly
- Add enableFIPSCheck feature flag and fipsScriptPath options in rego_query.go
- Add parallel processing with semaphore (max 3 concurrent checks)
- Implement result caching to avoid duplicate image checks
- Optimize extractor.go with de-duplication logic for FIPS fields
- Reorganize tests following Go conventions in internal/fipschecker/
- Add proper integration tests for single image and config workflows

Performance improvements:
- Parallel execution speeds up multiple unique image checks
- Caching eliminates redundant checks for duplicate images
- Semaphore prevents Docker daemon resource exhaustion
